File: tests/remediation/test_opening_identity.py
"""Identity lifecycle regression, including historical revision preservation.

No network, provider, filesystem migration or production mutation is performed.
"""
import copy
import json
from pathlib import Path
import sys
import unittest
from unittest.mock import patch

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
import acs_authoring as A
import acs_arch as AR
import acs_bim as B
import acs_workspace as W
import acs_understand as U
import acs_distance as DIST
import acs_relations as REL
import acs_opening_identity as OI


def fixture():
    return {"site": {"w": 10, "d": 10}, "floor_height": 4, "wall_h": 4, "wall_t": 0.2,
            "levels": [{"index": 0, "template": "g"}, {"index": 1, "template": "g"}],
            "floors": {"g": {"rooms": [{"id": "r", "rect": [0, 0, 10, 10],
                "doors": [{"edge": "N", "offset": 2, "width": 1, "height": 2.1},
                          {"edge": "S", "offset": 7, "width": 1, "height": 2.1}],
                "windows": [{"edge": "E", "offset": 2, "width": 1, "height": 1, "sill": 1},
                            {"edge": "W", "offset": 7, "width": 1, "height": 1, "sill": 1}]}]}}}


def room(m):
    return m["floors"]["g"]["rooms"][0]


def command(kind, target, parameters=None):
    return {"type": kind, "target_id": target, "parameters": parameters or {}}


def commit(p, cmd):
    v = A.validate_transaction(p, [cmd])
    result = A.commit_transaction(p, [cmd], confirm=v["transaction"]["confirmation_digest"],
                                  acknowledge_warnings=True, actor_id="test-user",
                                  created_at="2026-09-06T00:00:00Z")
    if not result["committed"]:
        raise AssertionError(result["issues"])
    return result["project"]


def snapshot():
    p = A.create_project(fixture())
    first = commit(p, command("DELETE_DOOR", "bld_0.g.r.door_0"))
    added = commit(first, command("ADD_DOOR", "bld_0.g.r",
                                {"edge": "E", "offset": 4, "width": 1, "height": 2.1}))
    aid = room(added["model"])["doors"][-1]["id"]
    moved = commit(added, command("MOVE_DOOR", aid, {"edge": "E", "offset": 5}))
    undone = A.undo(moved)["project"]
    redone = A.redo(undone)["project"]
    loaded = A.load_project(json.loads(json.dumps(A.serialise_project(redone, True, True))))["project"]
    arch = AR.compile_architecture(loaded["model"])
    return {"model": loaded["model"], "hash": loaded["model_hash"],
            "history": loaded["history"], "audit": loaded["audit_log"],
            "arch": arch, "tree": W.project_tree(loaded, arch),
            "relationships": REL.build_relationships(loaded["model"]),
            "exchange": B.build_exchange(loaded)["exchange"],
            "deleted": A.resolve_target(loaded["model"], "bld_0.g.r.door_0"),
            "surviving": A.resolve_target(loaded["model"], "bld_0.g.r.door_1"),
            "new_id": aid}


class OpeningIdentityTests(unittest.TestCase):
    def test_reserved_projection_names_match_authoring_contract(self):
        self.assertEqual(OI.DERIVED_NS, A.DERIVED_NS)

    def test_deleted_target_cannot_retarget_survivor(self):
        for kind in ("DOOR", "WINDOW"):
            with self.subTest(kind=kind):
                p = commit(A.create_project(fixture()), command("DELETE_" + kind, "bld_0.g.r." + kind.lower() + "_0"))
                self.assertIsNone(A.resolve_target(p["model"], "bld_0.g.r." + kind.lower() + "_0")["kind"])
                self.assertEqual(A.resolve_target(p["model"], "bld_0.g.r." + kind.lower() + "_1")["opening_index"], 0)

    def test_migration_is_idempotent_and_geometry_unchanged(self):
        m = fixture()
        before = AR.compile_architecture(m)
        self.assertTrue(A.stabilise_opening_ids(m))
        self.assertEqual(A.stabilise_opening_ids(m), [])
        self.assertEqual(AR.compile_architecture(m), before)

    def test_preview_cancel_and_old_revision_are_unchanged(self):
        p = A.create_project(fixture())
        before = copy.deepcopy(p)
        A.preview_command(p["model"], command("DELETE_DOOR", "bld_0.g.r.door_0"))
        self.assertEqual(p, before)
        q = commit(p, command("DELETE_DOOR", "bld_0.g.r.door_0"))
        self.assertEqual(p, before)
        self.assertEqual(q["revision_models"][p["current_revision"]], before["model"])
        self.assertIn("_opening_identity", q["audit_log"][-1]["changed_paths"])

    def test_add_move_save_undo_redo_preserve_identity(self):
        out = snapshot()
        self.assertEqual(out["surviving"]["opening_index"], 0)
        self.assertIsNone(out["deleted"]["kind"])
        self.assertEqual(room(out["model"])["doors"][-1]["id"], out["new_id"])
        self.assertEqual(room(out["model"])["doors"][-1]["offset"], 5)
        self.assertTrue(any(x["opening_ref"] == out["new_id"] for x in out["arch"]["openings"]))
        self.assertTrue(any(x["canonical_id"] == out["new_id"] + "@0" for x in out["exchange"]["doors"]))

    def test_reordering_keeps_identity_and_invalidates_positional_alias(self):
        m = fixture()
        A.stabilise_opening_ids(m)
        room(m)["doors"].reverse()
        self.assertEqual(A.resolve_target(m, "bld_0.g.r.door_0")["opening_index"], 1)
        self.assertEqual(A.resolve_target(m, "bld_0.g.r.door_1@1")["opening_index"], 0)
        self.assertIsNone(A.resolve_target(m, "bld_0.g.r.door_1@99")["kind"])

    def test_duplicate_and_reserved_identities_rejected(self):
        for identity in ("bld_0.g.r.door_0", "bld_0.g.r", "bld_0.flr_0", "__proto__"):
            with self.subTest(identity=identity):
                m = fixture()
                room(m)["windows"][0]["id"] = identity
                self.assertIn("ID_COLLISION", [i["code"] for i in A.opening_identity_issues(m)])
                with self.assertRaises(ValueError):
                    A.stabilise_opening_ids(m)

    def test_invalid_identities_and_marker_fail_closed(self):
        for value in (None, "", [], 12, "door@0", "door\nscript", "x" * 257, "<svg/onload=1>"):
            m = fixture()
            room(m)["doors"][0]["id"] = value
            with self.subTest(value=value), self.assertRaises(ValueError):
                A.stabilise_opening_ids(m)
        for state in ({"schema": "future", "next": 0}, {"schema": A.OPENING_IDENTITY_SCHEMA, "next": True}):
            m = fixture()
            m["_opening_identity"] = state
            self.assertTrue(A.opening_identity_issues(m))

    def test_missing_identity_in_migrated_model_is_not_regenerated(self):
        m = fixture()
        A.stabilise_opening_ids(m)
        del room(m)["doors"][0]["id"]
        before = copy.deepcopy(m)
        with self.assertRaises(ValueError):
            A.stabilise_opening_ids(m)
        self.assertEqual(m, before)

    def test_new_ids_are_not_reused_after_delete(self):
        p = A.create_project(fixture())
        add = command("ADD_WINDOW", "bld_0.g.r", {"edge": "E", "offset": 5, "width": 1})
        p = commit(p, add)
        oid = room(p["model"])["windows"][-1]["id"]
        p = commit(p, command("DELETE_WINDOW", oid))
        p = commit(p, add)
        self.assertNotEqual(room(p["model"])["windows"][-1]["id"], oid)

    def test_source_lock_cannot_be_bypassed_with_level_suffix(self):
        m = fixture()
        A.stabilise_opening_ids(m)
        m["_authoring_locks"] = {"bld_0.g.r.door_1": {"reason": "USER_LOCKED"}}
        result = A.preview_command(m, command("MOVE_DOOR", "bld_0.g.r.door_1@0", {"offset": 6}))
        self.assertFalse(result["valid"])
        self.assertIn("TARGET_LOCKED", [i["code"] for i in result["issues"]])

    def test_workspace_host_matches_selected_opening(self):
        p = A.create_project(fixture())
        A.stabilise_opening_ids(p["model"])
        arch = AR.compile_architecture(p["model"])
        oid = "bld_0.g.r.door_1"
        expected = next(o["host_wall_id"] for o in arch["openings"] if o["opening_ref"] == oid)
        rel = W._relationships(p["model"], A.resolve_target(p["model"], oid), arch, "bld_0")
        self.assertEqual(next(r["target_id"] for r in rel if r["relation"] == "HOSTED_BY_WALL"), expected)

    def test_new_generation_admission_assigns_ids_without_changing_unknown_thickness(self):
        m = fixture()
        m["wall_t"] = 0
        out = U.validate(m)
        self.assertEqual(out["wall_t"], 0)  # separate known uncertainty defect
        self.assertTrue(all(o.get("id") for k in ("doors", "windows") for o in room(out)[k]))

    def test_distance_uses_identity_after_reorder(self):
        m = fixture()
        A.stabilise_opening_ids(m)
        room(m)["doors"].reverse()
        idx = DIST._rooms_index(m, "bld_0")
        self.assertEqual(DIST._via_door("bld_0.g.r.door_0", idx), ("bld_0.g.r", 1))
        room(m)["doors"].pop()
        self.assertEqual(DIST._via_door("bld_0.g.r.door_0", idx), (None, None))

    def test_missing_distance_reference_is_not_a_legacy_door(self):
        rooms = {"r": {"doors": [{"edge": "N"}]}}
        for via in (None, "", 0):
            self.assertEqual(DIST._via_door(via, rooms), (None, None))

    def test_import_and_export_reject_broken_migrated_identity(self):
        m = fixture()
        A.stabilise_opening_ids(m)
        del room(m)["doors"][0]["id"]
        p = A.create_project(m)
        loaded = A.load_project(A.serialise_project(p))
        self.assertFalse(loaded["valid"])
        self.assertIsNone(loaded["project"])
        exported = B.build_exchange(p)
        self.assertFalse(exported["valid"])
        self.assertIsNone(exported["exchange"])

    def test_ai_edit_cannot_drop_identity_version_or_records(self):
        for omit in ("marker", "id"):
            m = fixture()
            before = copy.deepcopy(m)
            response = copy.deepcopy(m)
            A.stabilise_opening_ids(response)
            if omit == "marker":
                del response["_opening_identity"]
            else:
                del room(response)["doors"][0]["id"]
            with self.subTest(omit=omit), patch.object(U, "call_llm", return_value=json.dumps(response)):
                with self.assertRaises((U.E.AcsApiError, ValueError)):
                    U.apply_notes(m, [])
            self.assertEqual(m, before)


if __name__ == "__main__":
    if "--snapshot" in sys.argv:
        print(json.dumps(snapshot(), ensure_ascii=False, sort_keys=True))
    else:
        unittest.main(verbosity=2)

File: tests/remediation/test_opening_identity.js
/* Runs against the shipped JS in Node and in Playwright-managed Chromium. */
(function(){
  let passed=0,failed=0;
  const check=(name,ok)=>{if(ok){passed++;console.log('  ✓ '+name);}else{failed++;console.log('  ✗ '+name);}};
  const fixture=()=>({site:{w:10,d:10},floor_height:4,wall_h:4,wall_t:0.2,
    levels:[{index:0,template:'g'},{index:1,template:'g'}],floors:{g:{rooms:[
      {id:'r',rect:[0,0,10,10],doors:[{edge:'N',offset:2,width:1,height:2.1},
        {edge:'S',offset:7,width:1,height:2.1}],windows:[
        {edge:'E',offset:2,width:1,height:1,sill:1},{edge:'W',offset:7,width:1,height:1,sill:1}]}]}}});
  const room=m=>m.floors.g.rooms[0];
  const command=(type,target,parameters)=>({type:type,target_id:target,parameters:parameters||{}});
  function commit(p,cmd){
    const v=auValidateTransaction(p,[cmd]);
    const r=auCommitTransaction(p,[cmd],{confirm:v.transaction.confirmation_digest,
      acknowledge_warnings:true,actor_id:'test-user',created_at:'2026-09-06T00:00:00Z'});
    if(!r.committed) throw new Error(JSON.stringify(r.issues));
    return r.project;
  }
  for(const kind of ['DOOR','WINDOW']){
    const p=commit(auCreateProject(fixture()),command('DELETE_'+kind,'bld_0.g.r.'+kind.toLowerCase()+'_0'));
    check(kind+' deleted reference is not retargeted',auResolveTarget(p.model,'bld_0.g.r.'+kind.toLowerCase()+'_0').kind===null);
    check(kind+' survivor keeps its identity',auResolveTarget(p.model,'bld_0.g.r.'+kind.toLowerCase()+'_1').opening_index===0);
  }
  const m=fixture(), geom=JSON.stringify(compileArchitecture(m));
  check('migration records paths',auStabiliseOpeningIds(m).length>0);
  check('migration is idempotent',auStabiliseOpeningIds(m).length===0);
  check('migration leaves compiled geometry unchanged',JSON.stringify(compileArchitecture(m))===geom);
  room(m).doors.reverse();
  check('reorder preserves source identity',auResolveTarget(m,'bld_0.g.r.door_0').opening_index===1);
  check('valid template instance resolves',auResolveTarget(m,'bld_0.g.r.door_1@1').opening_index===0);
  check('invalid template instance rejected',auResolveTarget(m,'bld_0.g.r.door_1@99').kind===null);
  check('distance lookup survives reorder',_viaDoor('bld_0.g.r.door_0',_dsRooms(m,'bld_0'))[1]===1);
  for(const id of ['bld_0.g.r.door_0','bld_0.g.r','bld_0.flr_0','__proto__']){
    const bad=fixture();room(bad).windows[0].id=id;
    check('duplicate/reserved rejected: '+id,auOpeningIdentityIssues(bad).some(i=>i.code==='ID_COLLISION'));
  }
  for(const id of [null,'',[],12,'door@0','door\nscript','x'.repeat(257),'<svg/onload=1>']){
    const bad=fixture();room(bad).doors[0].id=id;
    check('invalid identity rejected',auOpeningIdentityIssues(bad).length>0);
  }
  const missing=fixture();auStabiliseOpeningIds(missing);delete room(missing).doors[0].id;
  check('missing migrated ID rejected',auOpeningIdentityIssues(missing).length>0);
  const invalidProject=auCreateProject(missing);
  check('import rejects broken migrated identity',!auLoadProject(auSerialiseProject(invalidProject)).valid);
  check('export rejects broken migrated identity',!bxBuildExchange(invalidProject).valid);
  check('missing distance reference is not a legacy door',
    _viaDoor(undefined,{r:{doors:[{edge:'N'}]}})[0]===null);
  const locked=fixture();auStabiliseOpeningIds(locked);
  locked._authoring_locks={'bld_0.g.r.door_1':{reason:'USER_LOCKED'}};
  const lockResult=auPreviewCommand(locked,command('MOVE_DOOR','bld_0.g.r.door_1@0',{offset:6}));
  check('level suffix cannot bypass source lock',!lockResult.valid&&lockResult.issues.some(i=>i.code==='TARGET_LOCKED'));

  const p=auCreateProject(fixture()), before=JSON.stringify(p);
  const first=commit(p,command('DELETE_DOOR','bld_0.g.r.door_0'));
  check('original project/history preserved',JSON.stringify(p)===before);
  check('migration belongs to same revision',first.audit_log[0].changed_paths.includes('_opening_identity'));
  const added=commit(first,command('ADD_DOOR','bld_0.g.r',{edge:'E',offset:4,width:1,height:2.1}));
  const aid=room(added.model).doors.slice(-1)[0].id;
  const moved=commit(added,command('MOVE_DOOR',aid,{edge:'E',offset:5}));
  const undone=auUndo(moved).project, redone=auRedo(undone).project;
  const loaded=auLoadProject(JSON.parse(JSON.stringify(auSerialiseProject(redone,true,true)))).project;
  check('save/undo/redo retain new ID',room(loaded.model).doors.slice(-1)[0].id===aid);
  check('redo restores geometry',room(loaded.model).doors.slice(-1)[0].offset===5);
  const arch=compileArchitecture(loaded.model), exchange=bxBuildExchange(loaded).exchange;
  check('compiler retains source reference',arch.openings.some(o=>o.opening_ref===aid));
  check('exchange retains instance ID',exchange.doors.some(o=>o.canonical_id===aid+'@0'));
  const again=commit(commit(loaded,command('DELETE_DOOR',aid)),command('ADD_DOOR','bld_0.g.r',{edge:'E',offset:4,width:1,height:2.1}));
  check('new ID not reused after deletion',room(again.model).doors.slice(-1)[0].id!==aid);
  const relations=_wsRelationships(loaded.model,auResolveTarget(loaded.model,aid),arch,'bld_0');
  check('inspector resolves exact opening host',relations.find(r=>r.relation==='HOSTED_BY_WALL').target_id===
    arch.openings.find(o=>o.opening_ref===aid).host_wall_id);
  const out={model:loaded.model,hash:loaded.model_hash,history:loaded.history,audit:loaded.audit_log,
    arch:arch,tree:wsProjectTree(loaded,arch,null,'en'),exchange:exchange,
    relationships:buildRelationships(loaded.model),
    deleted:auResolveTarget(loaded.model,'bld_0.g.r.door_0'),
    surviving:auResolveTarget(loaded.model,'bld_0.g.r.door_1'),new_id:aid};
  if(typeof document==='undefined'&&process.env.ACS_OPENING_IDENTITY_PARITY)
    require('fs').writeFileSync(process.env.ACS_OPENING_IDENTITY_PARITY,JSON.stringify(out));
  console.log('OPENING IDENTITY: '+passed+' passed, '+failed+' failed');
  if(typeof document!=='undefined') console.log('OPENING IDENTITY: executed in browser');
  if(failed) process.exit(1);
})();
